File: Sources/ReminderApp/ExerciseCoach.swift
import AppKit
import Foundation
import ReminderCore
import ReminderUI

/// Drives one takeover's guided exercises: the session cursor, the tick
/// timer, which exercises are done or cancelled, and the one place its audio
/// comes from.
///
/// Created once per takeover and shared by every display's
/// `CriticalOverlayView`, so a cue is spoken once on a multi-monitor desk and
/// a tick on one display shows on all of them. The session itself is the
/// pure `ExerciseSession` from the core; this class only feeds it the clock
/// and acts on what it reports.
///
/// With the voice on, every phase is announced before it is timed: the
/// session is frozen at the phase start while the cue is spoken and released
/// when the synthesizer says it has finished, so "hold for 5 seconds" is
/// followed by five seconds, not by whatever was left after the sentence.
@MainActor
final class ExerciseCoach: ObservableObject {
    /// Done: ticked by hand (untimed) or finished by the coach (guided).
    @Published private(set) var completedExerciseIDs: Set<UUID>
    /// Guided exercises the user has said they are not doing this time.
    @Published private(set) var cancelledExerciseIDs: Set<UUID>
    @Published private(set) var session: ExerciseSession?
    /// The exercise the session is coaching; `nil` once it has completed.
    @Published private(set) var activeExerciseID: UUID?
    /// The guided exercise whose Start is highlighted: the first neither
    /// done nor cancelled, so Space starts the obvious next thing.
    @Published private(set) var suggestedExerciseID: UUID?
    /// Republished on every tick, so views derive the countdown from
    /// `session` and this rather than keeping time of their own.
    @Published private(set) var now: Date

    let exercises: [Exercise]
    var hasGuidedExercises: Bool { exercises.contains(where: \.isGuided) }

    private let speech: SpeechCoaching?
    private let playsSounds: Bool
    private let clock: () -> Date
    private var timer: Timer?
    private var spokenCueCount = 0
    private var lastPhaseIndex: Int?
    private var sleepObserver: NSObjectProtocol?

    /// - Parameters:
    ///   - speech: `nil` when the voice coach is off; chimes still play and
    ///     phases start on time rather than after an announcement.
    ///   - clock: injected so a snapshot can freeze the coach mid-session.
    init(
        exercises: [Exercise],
        settings: ReminderCore.Settings,
        speech: SpeechCoaching?,
        clock: @escaping () -> Date = Date.init
    ) {
        self.exercises = exercises
        self.speech = speech
        self.playsSounds = settings.playsSound(for: .critical)
        self.clock = clock
        completedExerciseIDs = []
        cancelledExerciseIDs = []
        now = clock()
        suggestedExerciseID = Self.suggested(in: exercises, completed: [], cancelled: [])
    }

    // MARK: - Session control

    /// Coaches `id` from the top, replacing any session in progress.
    func start(_ id: UUID) {
        guard let exercise = exercises.first(where: { $0.id == id }),
              let timeline = ExerciseTimeline(exercise: exercise)
        else { return }
        speech?.stop()
        let startedAt = clock()
        session = ExerciseSession(timeline: timeline, startedAt: startedAt)
        activeExerciseID = id
        completedExerciseIDs.remove(id)
        cancelledExerciseIDs.remove(id)
        spokenCueCount = 0
        lastPhaseIndex = nil
        recomputeSuggested()
        observeSleep()
        tick()  // Announces "Get ready" straight away rather than a tick later.
        startTimer()
    }

    func startSuggested() {
        if let id = suggestedExerciseID { start(id) }
    }

    /// Space: pause or resume a session in progress, otherwise start the
    /// suggested exercise.
    func pauseResumeOrStart() {
        if let session, session.isLive {
            togglePause()
        } else {
            startSuggested()
        }
    }

    func togglePause() {
        guard var session, session.isLive else { return }
        let now = clock()
        session.togglePause(at: now)
        switch session.state {
        case .paused:
            speech?.stop()
            self.session = session
        case .running:
            self.session = session
            // A pause that cut the cue short leaves the phase untouched;
            // say it again so the person knows what they are resuming into.
            if speech != nil, session.isAtPhaseStart(at: now) {
                announceCurrentPhase()
            }
        default:
            self.session = session
        }
        tick()
    }

    /// Jumps to the next phase. Anything half-said is cut off so it cannot
    /// run into the next cue.
    func skip() {
        guard var session, session.isLive else { return }
        speech?.stop()
        session.skip(at: clock())
        if session.state == .announcing {
            // The announcement that was in progress is gone with its phase;
            // the tick announces the new one.
            session.finishAnnouncement(at: clock())
        }
        self.session = session
        tick()
    }

    /// Abandons the session; the exercise stays neither done nor cancelled.
    func stop() {
        speech?.stop()
        stopTimer()
        session = nil
        activeExerciseID = nil
        recomputeSuggested()
    }

    // MARK: - Ticks and cancels

    /// The tick box on an untimed exercise.
    func toggle(_ id: UUID) {
        if completedExerciseIDs.contains(id) {
            completedExerciseIDs.remove(id)
        } else {
            completedExerciseIDs.insert(id)
            if id == activeExerciseID { stop() }
        }
        recomputeSuggested()
    }

    /// "Not doing this one": dims the row and moves the suggestion on.
    /// Start on the row takes it back.
    func cancel(_ id: UUID) {
        if id == activeExerciseID { stop() }
        completedExerciseIDs.remove(id)
        cancelledExerciseIDs.insert(id)
        recomputeSuggested()
    }

    /// How the row for `id` should present itself; `nil` for an untimed
    /// exercise, which keeps its plain tick box.
    func rowState(for id: UUID) -> ExerciseRowCoachState? {
        guard let exercise = exercises.first(where: { $0.id == id }), exercise.isGuided else {
            return nil
        }
        if id == activeExerciseID, let session, let phase = session.phase(at: now) {
            let caption = "\(phase.title) · \(phase.label)"
            return .active(caption: session.state == .paused ? "Paused · \(caption)" : caption)
        }
        if completedExerciseIDs.contains(id) { return .completed }
        if cancelledExerciseIDs.contains(id) { return .cancelled }
        return id == suggestedExerciseID ? .suggested : .idle
    }

    /// Silences and stops everything. Done, Snooze and dismissing the
    /// takeover all end here.
    func shutDown() {
        speech?.stop()
        stopTimer()
        if let sleepObserver {
            NSWorkspace.shared.notificationCenter.removeObserver(sleepObserver)
            self.sleepObserver = nil
        }
    }

    // MARK: - Snapshots

    /// A coach frozen `offset` seconds into `active`'s session, with no timer
    /// and no voice, so the snapshot harness can render the panel in a known
    /// state.
    static func previewing(
        exercises: [Exercise],
        settings: ReminderCore.Settings,
        completed: Set<UUID>,
        cancelled: Set<UUID> = [],
        active: UUID,
        atOffset offset: TimeInterval,
        paused: Bool = false
    ) -> ExerciseCoach {
        let frozenNow = Date()
        let coach = ExerciseCoach(
            exercises: exercises, settings: settings, speech: nil, clock: { frozenNow }
        )
        coach.completedExerciseIDs = completed
        coach.cancelledExerciseIDs = cancelled
        if let exercise = exercises.first(where: { $0.id == active }),
           let timeline = ExerciseTimeline(exercise: exercise) {
            var session = ExerciseSession(
                timeline: timeline, startedAt: frozenNow.addingTimeInterval(-offset)
            )
            if paused { session.pause(at: frozenNow) }
            coach.session = session
            coach.activeExerciseID = active
        }
        coach.recomputeSuggested()
        return coach
    }

    // MARK: - The tick

    private func tick() {
        now = clock()
        guard var session else { return }

        let phaseIndex = session.timeline.phaseIndex(at: session.elapsed(at: now))
        let phaseChanged = phaseIndex != lastPhaseIndex
        if phaseChanged, lastPhaseIndex != nil, phaseIndex != nil, playsSounds {
            Sounds.play(named: "Tink")
        }
        lastPhaseIndex = phaseIndex

        if let speech {
            if phaseChanged, phaseIndex != nil, session.state == .running {
                // A new phase with the voice on: freeze at its start, say its
                // cue, and let the clock go when the sentence is over.
                session.beginAnnouncement(at: now)
                self.session = session
                spokenCueCount = session.cueCount(at: now)
                announceCurrentPhase()
                return
            }
            // Cues inside a phase (the "Three. Two. One." countdown) and the
            // closing line are spoken without stopping the clock. Only the
            // newest: after a stall the ones in between are stale, not a
            // backlog to read out.
            let cueCount = session.cueCount(at: now)
            if cueCount > spokenCueCount, session.state == .running {
                speech.speak(session.timeline.cues[cueCount - 1].text)
            }
            spokenCueCount = max(spokenCueCount, cueCount)
        }

        if session.markCompletedIfFinished(at: now) {
            self.session = session
            if let activeExerciseID {
                completedExerciseIDs.insert(activeExerciseID)
            }
            activeExerciseID = nil
            recomputeSuggested()
            if playsSounds { Sounds.play(named: "Glass") }
            stopTimer()
        }
    }

    /// Says the current phase's cue and releases the session when it has
    /// been said. The session may already be announcing (a fresh phase) or
    /// running at a phase start (a resume); either way it is frozen first so
    /// the timing is the same.
    private func announceCurrentPhase() {
        guard let speech, var session, let phase = session.phase(at: now) else { return }
        if session.state == .running {
            session.beginAnnouncement(at: now)
            self.session = session
        }
        let cue = ExerciseTimeline.cue(for: phase, exerciseName: session.timeline.exerciseName)
        speech.speak(cue) { [weak self] in
            self?.releaseAnnouncement(of: phase)
        }
        // If the synthesizer never reports back (no usable voice, say), the
        // hold must still start: nobody should be frozen at "Get ready".
        Task { [weak self] in
            try? await Task.sleep(nanoseconds: 8_000_000_000)
            self?.releaseAnnouncement(of: phase)
        }
    }

    /// Lets the clock run if the session is still frozen at the start of
    /// `phase`; a no-op once it has moved on, so a late release cannot
    /// touch a later phase's announcement.
    private func releaseAnnouncement(of phase: ExercisePhase) {
        guard var session, session.state == .announcing,
              session.phase(at: now) == phase
        else { return }
        session.finishAnnouncement(at: clock())
        self.session = session
        tick()
    }

    private func startTimer() {
        stopTimer()
        let timer = Timer(timeInterval: 0.2, repeats: true) { [weak self] _ in
            Task { @MainActor [weak self] in self?.tick() }
        }
        // Keep counting while a menu is open or a window is being dragged.
        RunLoop.main.add(timer, forMode: .common)
        self.timer = timer
    }

    private func stopTimer() {
        timer?.invalidate()
        timer = nil
    }

    /// A hold must not "complete" under a closed lid: pause when the Mac
    /// sleeps, and let the user resume when they are back in position.
    private func observeSleep() {
        guard sleepObserver == nil else { return }
        sleepObserver = NSWorkspace.shared.notificationCenter.addObserver(
            forName: NSWorkspace.willSleepNotification, object: nil, queue: .main
        ) { [weak self] _ in
            Task { @MainActor [weak self] in
                guard let self, var session = self.session,
                      session.state == .running || session.state == .announcing
                else { return }
                self.speech?.stop()
                session.pause(at: self.clock())
                self.session = session
            }
        }
    }

    private func recomputeSuggested() {
        suggestedExerciseID = Self.suggested(
            in: exercises, completed: completedExerciseIDs, cancelled: cancelledExerciseIDs
        )
    }

    private static func suggested(
        in exercises: [Exercise], completed: Set<UUID>, cancelled: Set<UUID>
    ) -> UUID? {
        exercises.first { $0.isGuided && !completed.contains($0.id) && !cancelled.contains($0.id) }?.id
    }
}

File: Windows/Pauselet.App/ExerciseCoach.cs
using System.Windows.Threading;
using NodaTime;
using Pauselet.Core;

namespace Pauselet.App;

/// <summary>
/// How a row in the takeover should present itself while the coach runs.
/// (Mirrors ExerciseRowCoachState in ReminderUI.)
/// </summary>
internal enum ExerciseRowCoachState
{
    /// <summary>Waiting its turn.</summary>
    Idle,
    /// <summary>The exercise whose Start is highlighted.</summary>
    Suggested,
    /// <summary>Being coached now.</summary>
    Active,
    Completed,
    /// <summary>The user said they are not doing this one.</summary>
    Cancelled,
}

/// <summary>
/// Drives one takeover's exercises: the session cursor, the tick timer,
/// which exercises are done or cancelled, and the one place its audio comes
/// from. (Mirrors ExerciseCoach.swift.)
///
/// Created once per takeover and shared by every display's overlay, so a cue
/// is spoken once on a multi-monitor desk and a finished exercise shows on
/// all of them. The session itself is the pure <see cref="ExerciseSession"/>
/// from the core; this class only feeds it the clock and acts on what it
/// reports.
///
/// A session coaches one exercise (Start on its row) or every exercise still
/// to do, in order, with the reminder's rest between them (Start all). Either
/// way the timeline is built once, up front, by the core.
///
/// With the voice on, every phase is announced before it is timed: the
/// session is frozen at the phase start while the cue is spoken and released
/// when the synthesizer says it has finished, so "hold for 5 seconds" is
/// followed by five seconds, not by whatever was left after the sentence.
/// </summary>
internal sealed class ExerciseCoach : IDisposable
{
    /// <summary>Finished by the coach.</summary>
    public HashSet<Guid> CompletedExerciseIds { get; } = [];
    /// <summary>Exercises the user has said they are not doing this time.</summary>
    public HashSet<Guid> CancelledExerciseIds { get; } = [];
    public ExerciseSession? Session { get; private set; }
    /// <summary>
    /// The exercise whose Start is highlighted: the first neither done nor
    /// cancelled, so the obvious next thing is one keypress away.
    /// </summary>
    public Guid? SuggestedExerciseId { get; private set; }
    /// <summary>
    /// Updated on every tick, so views derive the countdown from
    /// <see cref="Session"/> and this rather than keeping time of their own.
    /// </summary>
    public Instant Now { get; private set; }

    /// <summary>Raised on every tick and every state change, on the UI thread.</summary>
    public event Action? Changed;

    public IReadOnlyList<Exercise> Exercises { get; }
    /// <summary>The reminder's rest between exercises, used by Start all.</summary>
    public int RestBetweenExercisesSeconds { get; }

    /// <summary>
    /// The exercise the session is on right now; <c>null</c> between sessions
    /// and once one has completed.
    /// </summary>
    public Guid? ActiveExerciseId => Session?.PhaseAt(Now)?.ExerciseId;

    /// <summary>Neither done nor cancelled, in programme order: what Start all runs.</summary>
    public IReadOnlyList<Guid> RemainingExerciseIds => Exercises
        .Where(exercise =>
            !CompletedExerciseIds.Contains(exercise.Id)
            && !CancelledExerciseIds.Contains(exercise.Id))
        .Select(exercise => exercise.Id)
        .ToList();

    /// <summary>
    /// Start all earns its place once there are two or more left to run;
    /// with one, it is the same as Start.
    /// </summary>
    public bool CanStartAll => RemainingExerciseIds.Count > 1;

    private readonly ISpeechCoaching? _speech;
    private readonly bool _playsSounds;
    private readonly Func<Instant> _clock;
    private DispatcherTimer? _timer;
    private int _spokenCueCount;
    private int? _lastPhaseIndex;
    private DispatcherTimer? _announcementFallback;

    /// <param name="speech">
    /// <c>null</c> when the voice coach is off; chimes still play and phases
    /// start on time rather than after an announcement.
    /// </param>
    /// <param name="clock">Injected so a test can drive the coach without waiting.</param>
    public ExerciseCoach(
        IReadOnlyList<Exercise> exercises,
        int restBetweenExercisesSeconds,
        Settings settings,
        ISpeechCoaching? speech,
        Func<Instant>? clock = null)
    {
        Exercises = exercises;
        RestBetweenExercisesSeconds = restBetweenExercisesSeconds;
        _speech = speech;
        _playsSounds = settings.PlaysSound(Priority.Critical);
        _clock = clock ?? (() => SystemClock.Instance.GetCurrentInstant());
        Now = _clock();
        SuggestedExerciseId = Suggested(exercises, [], []);
    }

    // MARK: - Session control

    /// <summary>Coaches <paramref name="id"/> from the top, replacing any session in progress.</summary>
    public void Start(Guid id)
    {
        var exercise = Exercises.FirstOrDefault(candidate => candidate.Id == id);
        if (exercise is null || ExerciseTimeline.For(exercise) is not { } timeline) return;
        CompletedExerciseIds.Remove(id);
        CancelledExerciseIds.Remove(id);
        Run(timeline);
    }

    public void StartSuggested()
    {
        if (SuggestedExerciseId is { } id) Start(id);
    }

    /// <summary>
    /// Runs every exercise still to do, in order, with the reminder's rest
    /// between one and the next, replacing any session in progress.
    /// </summary>
    public void StartAll()
    {
        var remaining = RemainingExerciseIds;
        var queue = Exercises.Where(exercise => remaining.Contains(exercise.Id)).ToList();
        if (ExerciseTimeline.For(queue, RestBetweenExercisesSeconds) is not { } timeline) return;
        Run(timeline);
    }

    /// <param name="paused">
    /// A run handed over while paused (the exercise being coached was
    /// cancelled mid-pause) stays paused; nothing is said until the user
    /// resumes.
    /// </param>
    private void Run(ExerciseTimeline timeline, bool paused = false)
    {
        _speech?.Stop();
        StopAnnouncementFallback();
        var fresh = new ExerciseSession(timeline, _clock());
        if (paused) fresh.Pause(_clock());
        Session = fresh;
        _spokenCueCount = 0;
        _lastPhaseIndex = null;
        RecomputeSuggested();
        Tick();  // Announces "Get ready" straight away rather than a tick later.
        StartTimer();
    }

    /// <summary>
    /// Space: pause or resume a session in progress, otherwise start the
    /// suggested exercise.
    /// </summary>
    public void PauseResumeOrStart()
    {
        if (Session is { IsLive: true }) TogglePause();
        else StartSuggested();
    }

    public void TogglePause()
    {
        if (Session is not { IsLive: true } session) return;
        var now = _clock();
        session.TogglePause(now);
        if (session.State == ExerciseSession.SessionState.Paused)
        {
            _speech?.Stop();
        }
        else if (session.State == ExerciseSession.SessionState.Running
            && _speech is not null && session.IsAtPhaseStart(now))
        {
            // A pause that cut the cue short leaves the phase untouched; say
            // it again so the person knows what they are resuming into.
            AnnounceCurrentPhase();
        }
        Tick();
    }

    /// <summary>
    /// Jumps to the next phase. Anything half-said is cut off so it cannot run
    /// into the next cue.
    /// </summary>
    public void Skip() => Move((session, now) => session.Skip(now));

    private void Move(Action<ExerciseSession, Instant> reposition)
    {
        if (Session is not { IsLive: true } session) return;
        _speech?.Stop();
        var now = _clock();
        reposition(session, now);
        if (session.State == ExerciseSession.SessionState.Announcing)
        {
            // The announcement that was in progress is gone with its phase;
            // the tick announces the new one.
            session.FinishAnnouncement(now);
        }
        Tick();
    }

    /// <summary>Abandons the session; the exercise it was on stays neither done nor cancelled.</summary>
    public void Stop()
    {
        _speech?.Stop();
        StopTimer();
        // The phase whose announcement it was waiting on is gone with the
        // session, so the fallback has nothing left to release.
        StopAnnouncementFallback();
        Session = null;
        RecomputeSuggested();
        Changed?.Invoke();
    }

    // MARK: - Cancels

    /// <summary>
    /// "Not doing this one": dims the row and moves the suggestion on. Start on
    /// the row takes it back. An exercise still to come in the run in progress
    /// is taken out of it; the one being coached hands over to the next, or
    /// ends the session when it was the last (or on its own).
    /// </summary>
    public void Cancel(Guid id)
    {
        Now = _clock();  // Judge "being coached" by the clock, not the last tick.
        CompletedExerciseIds.Remove(id);
        CancelledExerciseIds.Add(id);
        RecomputeSuggested();
        if (Session is { IsLive: true } session
            && session.Timeline.Entries.FirstOrDefault(entry => entry.Id == id) is { } dropped
            && session.Elapsed(Now) < dropped.End)
        {
            DropFromRun(dropped, session);  // Raises Changed itself.
            return;
        }
        Changed?.Invoke();
    }

    /// <summary>
    /// Takes a cancelled exercise out of the run in progress. The timeline is
    /// rebuilt without it, which is what keeps the coaching honest: the
    /// exercise is never announced or run, its row never lights up, and the
    /// next lead-in signs off the exercise that actually finished before it.
    /// One still ahead of the cursor leaves the session exactly where it was,
    /// in the state it was in; the one being coached hands over to whatever
    /// follows it, from its lead-in, or ends the session when nothing does.
    /// (Mirrors dropFromRun in ExerciseCoach.swift.)
    /// </summary>
    private void DropFromRun(ExerciseTimeline.Entry dropped, ExerciseSession old)
    {
        var entries = old.Timeline.Entries;
        var isActive = old.PhaseAt(Now)?.ExerciseId == dropped.Id;
        var kept = isActive
            ? entries.SkipWhile(entry => entry.Id != dropped.Id).Skip(1).ToList()
            : entries.Where(entry => entry.Id != dropped.Id).ToList();
        if (isActive && kept.Count == 0)
        {
            // Nothing follows. A run of several is over — it completes, as it
            // would have a moment later; an exercise on its own just stops.
            if (entries.Count > 1)
            {
                Move((session, now) => session.Jump(session.Timeline.TotalDuration, now));
            }
            else
            {
                Stop();
            }
            return;
        }
        var queue = kept
            .Select(entry => Exercises.FirstOrDefault(exercise => exercise.Id == entry.Id))
            .OfType<Exercise>()
            .ToList();
        if (ExerciseTimeline.For(queue, RestBetweenExercisesSeconds) is not { } timeline)
        {
            Stop();
            return;
        }
        if (isActive)
        {
            Run(timeline, paused: old.State == ExerciseSession.SessionState.Paused);
            return;
        }
        // Everything before the cursor is unchanged — same phases, same cues —
        // so the session carries on from the same offset. An announcement in
        // flight still matches its phase and is released as normal when the
        // synthesizer reports back.
        var fresh = new ExerciseSession(timeline, Now);
        fresh.Jump(old.Elapsed(Now), Now);
        switch (old.State)
        {
            case ExerciseSession.SessionState.Paused: fresh.Pause(Now); break;
            case ExerciseSession.SessionState.Announcing: fresh.BeginAnnouncement(Now); break;
        }
        Session = fresh;
        Changed?.Invoke();
    }

    /// <summary>How the row for <paramref name="id"/> should present itself.</summary>
    public ExerciseRowCoachState RowState(Guid id)
    {
        if (Session?.PhaseAt(Now) is { } phase && phase.ExerciseId == id)
        {
            return ExerciseRowCoachState.Active;
        }
        if (CompletedExerciseIds.Contains(id)) return ExerciseRowCoachState.Completed;
        if (CancelledExerciseIds.Contains(id)) return ExerciseRowCoachState.Cancelled;
        return id == SuggestedExerciseId
            ? ExerciseRowCoachState.Suggested
            : ExerciseRowCoachState.Idle;
    }

    /// <summary>
    /// The caption for the active row: "Set 1 · Rep 3 · Hold", prefixed with
    /// "Paused" when it is. <c>null</c> when the row is not the active one.
    /// </summary>
    public string? RowCaption(Guid id)
    {
        if (id != ActiveExerciseId || Session is not { } session) return null;
        if (session.PhaseAt(Now) is not { } phase) return null;
        var caption = $"{phase.Title} · {phase.Label}";
        return session.State == ExerciseSession.SessionState.Paused
            ? $"Paused · {caption}"
            : caption;
    }

    /// <summary>
    /// Silences and stops everything. Done, Snooze and dismissing the takeover
    /// all end here.
    /// </summary>
    public void ShutDown()
    {
        _speech?.Stop();
        StopTimer();
        StopAnnouncementFallback();
        // The coach owns the synthesizer it was handed — it is built per
        // takeover — and it holds the default audio device until disposed.
        (_speech as IDisposable)?.Dispose();
    }

    // MARK: - The tick

    private void Tick()
    {
        Now = _clock();
        if (Session is not { } session)
        {
            Changed?.Invoke();
            return;
        }
        var timeline = session.Timeline;
        var elapsed = session.Elapsed(Now);

        // Exercises whose last set has run are done on every display, with
        // the finishing chime, and the suggestion moves on. A cancelled one
        // that was jumped over is left as cancelled.
        var phaseIndex = timeline.PhaseIndexAt(elapsed);
        var phaseChanged = phaseIndex != _lastPhaseIndex;
        // An exercise can only end at a phase boundary, so the sweep for
        // finished ones is skipped on the other four ticks a second.
        var finished = phaseChanged
            ? timeline.FinishedExerciseIds(elapsed)
                .Where(id => !CompletedExerciseIds.Contains(id) && !CancelledExerciseIds.Contains(id))
                .ToList()
            : [];
        if (finished.Count > 0)
        {
            CompletedExerciseIds.UnionWith(finished);
            RecomputeSuggested();
            if (_playsSounds) Sounds.Play("Glass");
        }

        // A phase boundary that is also an exercise boundary has had its chime.
        if (phaseChanged && _lastPhaseIndex is not null && phaseIndex is not null
            && finished.Count == 0 && _playsSounds)
        {
            Sounds.Play("Tink");
        }
        _lastPhaseIndex = phaseIndex;

        if (_speech is { } speech)
        {
            if (phaseChanged && phaseIndex is not null
                && session.State == ExerciseSession.SessionState.Running)
            {
                // A new phase with the voice on: freeze at its start, say its
                // cue, and let the clock go when the sentence is over.
                session.BeginAnnouncement(Now);
                _spokenCueCount = session.CueCount(Now);
                AnnounceCurrentPhase();
                Changed?.Invoke();
                return;
            }
            // Cues inside a phase (the "Three. Two. One." countdown) and the
            // closing line are spoken without stopping the clock. Only the
            // newest: after a stall the ones in between are stale, not a
            // backlog to read out.
            var cueCount = session.CueCount(Now);
            if (cueCount > _spokenCueCount
                && session.State == ExerciseSession.SessionState.Running)
            {
                speech.Speak(timeline.Cues[cueCount - 1].Text);
            }
            _spokenCueCount = Math.Max(_spokenCueCount, cueCount);
        }

        if (session.MarkCompletedIfFinished(Now))
        {
            StopTimer();
        }
        Changed?.Invoke();
    }

    /// <summary>
    /// Says the current phase's cue and releases the session when it has been
    /// said. The session may already be announcing (a fresh phase) or running
    /// at a phase start (a resume); either way it is frozen first so the
    /// timing is the same.
    /// </summary>
    private void AnnounceCurrentPhase()
    {
        if (_speech is not { } speech || Session is not { } session) return;
        if (session.PhaseAt(Now) is not { } phase) return;
        if (session.State == ExerciseSession.SessionState.Running)
        {
            session.BeginAnnouncement(Now);
        }
        speech.Speak(phase.Cue, () => ReleaseAnnouncement(phase));
        // If the synthesizer never reports back (no usable voice, say), the
        // hold must still start: nobody should be frozen at "Get ready".
        StartAnnouncementFallback(phase);
    }

    /// <summary>
    /// Lets the clock run if the session is still frozen at the start of
    /// <paramref name="phase"/>; a no-op once it has moved on, so a late
    /// release cannot touch a later phase's announcement.
    /// </summary>
    private void ReleaseAnnouncement(ExercisePhase phase)
    {
        StopAnnouncementFallback();
        if (Session is not { } session
            || session.State != ExerciseSession.SessionState.Announcing
            || session.PhaseAt(Now) != phase)
        {
            return;
        }
        session.FinishAnnouncement(_clock());
        Tick();
    }

    private void StartAnnouncementFallback(ExercisePhase phase)
    {
        StopAnnouncementFallback();
        _announcementFallback = new DispatcherTimer
        {
            Interval = TimeSpan.FromSeconds(8),
        };
        _announcementFallback.Tick += (_, _) => ReleaseAnnouncement(phase);
        _announcementFallback.Start();
    }

    private void StopAnnouncementFallback()
    {
        _announcementFallback?.Stop();
        _announcementFallback = null;
    }

    private void StartTimer()
    {
        StopTimer();
        // DispatcherTimer keeps counting while a menu is open or a window is
        // being dragged, and its callbacks land on the UI thread the overlay
        // is drawn on.
        _timer = new DispatcherTimer(DispatcherPriority.Render)
        {
            Interval = TimeSpan.FromMilliseconds(200),
        };
        _timer.Tick += (_, _) => Tick();
        _timer.Start();
    }

    private void StopTimer()
    {
        _timer?.Stop();
        _timer = null;
    }

    /// <summary>
    /// A hold must not "complete" while the PC is asleep or locked: pause, and
    /// let the user resume when they are back in position. The Mac's
    /// pause-on-sleep and iOS's pause-on-backgrounding, in the shape Windows
    /// reports it.
    /// </summary>
    public void SuspendForAbsence()
    {
        if (Session is not { } session) return;
        if (session.State != ExerciseSession.SessionState.Running
            && session.State != ExerciseSession.SessionState.Announcing)
        {
            return;
        }
        _speech?.Stop();
        session.Pause(_clock());
        Changed?.Invoke();
    }

    private void RecomputeSuggested() =>
        SuggestedExerciseId = Suggested(Exercises, CompletedExerciseIds, CancelledExerciseIds);

    private static Guid? Suggested(
        IReadOnlyList<Exercise> exercises, HashSet<Guid> completed, HashSet<Guid> cancelled) =>
        exercises.FirstOrDefault(exercise =>
            !completed.Contains(exercise.Id)
            && !cancelled.Contains(exercise.Id))?.Id;

    public void Dispose() => ShutDown();
}
